pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java

```diff
@@ -35,6 +35,14 @@
  */
 public interface AuthorizationProvider extends Closeable {
 
+    /**
+     * Check if specified role is a super user
+     * @param role the role to check
+     * @return a CompletableFuture containing a boolean in which true means the role is a super user
+     * and false if it is not
+     */
+    CompletableFuture<Boolean> isSuperUser(String role);
```

Comment:
this looks like breaking binary compatibility, if a user has a customized authorization provider.

can we try to make it a default method?

```
default CompletableFuture<Boolean> isSuperUser(String role) {
    return Futures.value(false);
}
```

Reply:
hmm, what you are saying here is this is a semantic breaking change?  

if that's case, can we still maintain semantic BC before breaking it? for example, you can deprecate the superUserRoles setting, but not change the semantic. so if superUserRoles is set, use the setting and maintain the BC behavior; if superUserRoles is not set, then use the new behavior. 

If you don't want to maintain the semantic BC, then you should highlight the BC changes somewhere and document what users should do when upgrading to 2.3.0.

Reply:
The problem here is that the current behavior is based on the list of super-user roles that comes from the config file. Having the `default` method in the interface will not help because it doesn't have access to the config file. 

At the same time, this is something that should have been added in the authorization plugin interface when we created. At this point, since I don't know of any alternative concrete implementations, I think it's better to fix the interface and document the breaking change in the release notes.

Reply:
@sijie The reason its hard to maintain backwards compatibility and have a default "isSuperUser" methods is because there is no good way to get the config "superUserRoles" from the config file in a default method.  I not sure how we can determine the location of the broker config file location from a inside a default method.

We can should definitely document this change.  Do we have a standard way of documenting BC breaking changes?  Should we just have comments in the PR? or a document for the release?

Reply:
> deprecate the setting in this release

This point here is not to deprecate the setting in `broker.conf`, that will still be the base of the default implementation of `isSuperUser()` method. Rather, we should allow to customize the authorization in all aspects, including the super-user status.

Reply:
> we should allow to customize the authorization in all aspects, including the super-user status.

okay. but my proposal allows customization and maintains BC, no?

Reply:
Adding the default method was indeed the first though (before Jerry started working on this), but I don't see how to get the ref of `ServiceConfiguration` from within the default method. Am I missing a different option?

Reply:
Sure, that would be a good option too. Personally I'd prefer to take the hit now and  "fix" the interface with documentation of the breaking change, but I'm fine either way.

Reply:
I am okay with taking the hit now. but we need to agree on an approach and have a mechanism to ensure documentation is in place for breaking changes. 

because people would quickly forget adding documentation after hitting the merge button, and no one is going to add this piece of documentation. after we release 2.3.0, community users will start complaining and ask questions. this has been seen many times in different BC related pull requests.

Reply:
Completely agree with that. We need to define a process around that. An option would be to require a PIP for each change that break compatibility (even in minimal/trivial ways).

Additionally we could maintain a wiki page with all breaking changes. The lists will have to be update when the PRs are merged and will serve as the base to compile the release notes.

Reply:
The purpose of requiring a PIP will be to attract more attention from dev community to have a change to speak up, since a PR discussion might otherwise be missed in the overall noise.

Reply:
yes a PIP is required for any breaking changes. that has been done in other communities, BP for bookkeeper, KIP for Kafka and FLIP for Flink.

Reply:
@merlimat any updates on the above?